### PR TITLE
[Cache] Add Redis Relay support

### PR DESCRIPTION
**Pull request by ostrolucky**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? |no
| Tickets       |
| License       | MIT
| Doc PR        | 

This PR adds support for [Relay](https://relay.so/), a next-gen Redis client written in C by the makers of PhpRedis and Predis. It’s built as a drop-in replacement for PhpRedis with a backwards compatible interface for easy adoption. Relay is significantly faster than existing clients by leveraging Redis 6's client-side-caching.

While Relay is still on 0.x (pending the addition of cluster support in a few weeks for a 1.0 tag), it’s interface is stable and it’s heavily used in production deployments.

Similarly, I've also added Relay support to most popular symfony redis bundle, [snc/redis-bundle](https://github.com/snc/SncRedisBundle/pull/688). But to be able to support these new Redis instances in Symfony components as a Cache, Lock and so on, support had to be added here as well.

Since method and constant declarations are compatible with  PhpRedis, I've opted into reusing most of the code instead of creating completely new adapters, similarly as was done in case of Predis+PhpRedis. At the same time, I made it a goal not having to have PhpRedis installed in case somebody wishes to use Relay, which explains things like conditional fetch of value constants.

### CODE REVIEW

**Comment by tillkruss** [comment]
Ping @michael-grunder.

File: src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php
Comment (tillkruss) [inline src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php]: Does it make sense to not import `Relay\Exception` for readability? Same question in a few other places. I'm not familiar with he Symfony best practices.

Comment (ostrolucky) [reply src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php]: I don't have preference on this, let's see if maintainers agree

Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php]: let's inline I agree, same as we did for `\Predis\ClientInterface`

File: src/Symfony/Component/Cache/CHANGELOG.md
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/CHANGELOG.md]: missing space at start of line ;)

File: src/Symfony/Component/Cache/Traits/RedisTrait.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: `Redis Sentinel support requires one of "predis/predis", "ext-redis >= 5.2" or "ext-relay": "%s".`

File: src/Symfony/Component/Cache/Traits/RedisTrait.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: `$class = $params['class'] ?? match (true) {`

File: src/Symfony/Component/Cache/Traits/RedisTrait.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: and neither

File: src/Symfony/Component/Cache/Traits/RedisTrait.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: `$isRedisExt`?

File: src/Symfony/Component/Cache/Traits/RedisTrait.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: Relay::OPT_TCP_KEEPALIVE

File: src/Symfony/Component/Lock/CHANGELOG.md
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/CHANGELOG.md]: missing space at start of line

File: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php]: extension_loaded would work better with opcache

File: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php]: extension_loaded

File: src/Symfony/Component/Semaphore/CHANGELOG.md
Comment (nicolas-grekas) [inline src/Symfony/Component/Semaphore/CHANGELOG.md]: missing space

File: src/Symfony/Component/VarDumper/Caster/RedisCaster.php
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/RedisCaster.php]: Aren't `Relay::OPT_PREFIX` & co. available?

File: src/Symfony/Component/VarDumper/Tests/Caster/RedisCasterTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Tests/Caster/RedisCasterTest.php]: better figure out away to run only the redis tests when relay isn't available, and vice versa

File: .github/workflows/integration-tests.yml
Comment (nicolas-grekas) [inline .github/workflows/integration-tests.yml:148]: what is this for?

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for working on this!
What's the plan for cluser/array clients?
For the proxy, it'd be great to amend RedisProxiesTest.php to sync it if needed.

File: src/Symfony/Component/VarDumper/Caster/RedisCaster.php
Comment (tillkruss) [reply src/Symfony/Component/VarDumper/Caster/RedisCaster.php]: They are: https://builds.r2.relay.so/dev/relay.stub.php

File: .github/workflows/integration-tests.yml
Comment (tillkruss) [reply .github/workflows/integration-tests.yml:148]: It's for telemetry, we use it to identify installations anonymously and what features they use, how long they run. It can be omitted. We are working on getting Relay into `shivammathur/setup-php`.

**Comment by tillkruss** [comment]
> What's the plan for cluser/array clients?

We're actively working on Cluster support, it will go into the 1.0 release in a few weeks. I'm not sure about `RedisArray`, I'll talk to @michael-grunder on whether we want to port that over, since there are about 2 production installs AFAIK.

**Comment by michael-grunder** [comment]
My preference would be to not replicate `RedisArray`, as we kind of want to deprecate the functionality even in PhpRedis.  That said, we could implement the API if there is sufficient demand.  It's much simpler than Cluster.

As @tillkruss said, we're working on Cluster support now.  Feel free to mention any quality-of-life features you'd like to see.

**Comment by nicolas-grekas** [comment]
Thanks for the hints. Another thing I'm wondering about is ext-redis v6: I'm seeing active dev on it. What's the plan related to that + relay?

**Comment by michael-grunder** [comment]
For now, I'm just making sure that the two extensions remain compatible.

File: .github/workflows/integration-tests.yml
Comment (ostrolucky) [reply .github/workflows/integration-tests.yml:148]: Let me know if I should remove this ;)

Comment (tillkruss) [reply .github/workflows/integration-tests.yml:148]: You can keep it, once the `shivammathur/setup-php` is merged it comes unnecessary.

**Comment by ostrolucky** [comment]
All feedback addressed

File: .github/workflows/integration-tests.yml
Comment (nicolas-grekas) [reply .github/workflows/integration-tests.yml:148]: I'd remove it then, this will reduce back and forth on this file

File: src/Symfony/Component/Cache/Tests/Traits/RedisProxiesTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Tests/Traits/RedisProxiesTest.php:53]: let's move the root `@requires` on the class to the corresponding methods?

File: src/Symfony/Component/Cache/Tests/Traits/RedisProxiesTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Tests/Traits/RedisProxiesTest.php]: this should reflect the `Relay` class, not the proxy

File: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php]: `\Relay\Exception` everywhere instead

**Comment by nicolas-grekas** [comment]
One thing I'm wondering is: did you consider keeping the same class names as phpredis?
Because if there are no differences API-wise, why not?
Couldn't relay be released as phpredis v5.4?

**Comment by ostrolucky** [comment]
Feedback addressed. But we have to add the random uuid back, ok @nicolas-grekas ?
```
1) Symfony\Component\Cache\Tests\Adapter\PredisAdapterSentinelTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

2) Symfony\Component\Cache\Tests\Adapter\PredisAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

3) Symfony\Component\Cache\Tests\Adapter\PredisClusterAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

4) Symfony\Component\Cache\Tests\Adapter\PredisRedisClusterAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

5) Symfony\Component\Cache\Tests\Adapter\PredisTagAwareAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

6) Symfony\Component\Cache\Tests\Adapter\PredisTagAwareClusterAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

7) Symfony\Component\Cache\Tests\Adapter\ProxyAdapterAndRedisAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

8) Symfony\Component\Cache\Tests\Adapter\RedisAdapterSentinelTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

9) Symfony\Component\Cache\Tests\Adapter\RedisAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

10) Symfony\Component\Cache\Tests\Adapter\RedisArrayAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

11) Symfony\Component\Cache\Tests\Adapter\RedisClusterAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

12) Symfony\Component\Cache\Tests\Adapter\RedisTagAwareAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

13) Symfony\Component\Cache\Tests\Adapter\RedisTagAwareArrayAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

14) Symfony\Component\Cache\Tests\Adapter\RedisTagAwareClusterAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

15) Symfony\Component\Cache\Tests\Adapter\RelayAdapterSentinelTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0

16) Symfony\Component\Cache\Tests\Adapter\RelayAdapterTest::testClearWithPrefix
PHPUnit\Framework\Exception: PHP Warning:  PHP Startup: Relay is using the default binary ID in Unknown on line 0
```

**Comment by nicolas-grekas** [comment]
> But we have to add the random uuid back

sure, let's just not forget to remove when possible

**Comment by ostrolucky** [comment]
done

File: .github/workflows/integration-tests.yml
Comment (tillkruss) [reply .github/workflows/integration-tests.yml:148]: Support was merged, but it's not released yet:

https://github.com/shivammathur/setup-php/issues/682#issuecomment-1381168907

**Comment by ostrolucky** [comment]
I've rebased the branch to solve the conflicts. Can we expect the merge soon? I promise I'll create a PR with migration to shivammathur/setup-php when they release the version ;)

**Comment by nicolas-grekas** [comment]
@michael-grunder @tillkruss any hints about [my previous comment](https://github.com/symfony/symfony/pull/48930#issuecomment-1381505142)? Why a new set of symbols? Why not release as phpredis v5.4?

File: src/Symfony/Component/Cache/Tests/Adapter/RelayAdapterTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Tests/Adapter/RelayAdapterTest.php]: let's move this to `@requires extension relay`

**Review by nicolas-grekas — APPROVED** [review]
LGTM thanks!

**Comment by tillkruss** [comment]
> One thing I'm wondering is: did you consider keeping the same class names as phpredis?

We did, but that would mean PhpRedis no Relay can't be installed at the same time and easily switched between. 

> Because if there are no differences API-wise, why not?

We did initially in 2020 consider this, but leveraging server-assisted client-side caching required starting from scratch, hence a separate name, new codebase, etc. 

PhpRedis has a lot of legacy junk that we didn't want to bring over. 

> Couldn't relay be released as phpredis v5.4?

People don't like change, PhpRedis will remain FOSS and we'll suggest users to move to Relay once 1.0 is out. 

Predis and PhpRedis will remain maintained of course, but who doesn't want a more modern, faster client ✌️

**Comment by nicolas-grekas** [comment]
(rebase needed :pray: + don't miss https://github.com/symfony/symfony/pull/48930#discussion_r1083447837)

**Comment by ostrolucky** [comment]
done

**Comment by ostrolucky** [comment]
Psalm found 1 relevant issue (missing \Relay\ in \Relay\Exception) so I fixed that. Not sure what is a desired fix for the other issues though. I think ideally, here
```php
if ($this->redis instanceof \Predis\ClientInterface) {
   $prefix = $this->redis->getOptions()->prefix ? $this->redis->getOptions()->prefix->getPrefix() : '';
  $prefixLen = \strlen($prefix ?? '');
}
```
before calling getPrefix, we should check if prefix is instanceof KeyPrefixProcessor.

And for other issues, ProxyHelper should be fixed to account for missing magic methods in parent. 

I think none of these are related to this PR.

**Comment by nicolas-grekas** [comment]
Thank you @ostrolucky.

**Comment by nicolas-grekas** [comment]
@tillkruss I'm sorry but I didn't manage to find the source code of the Relay extension. Can you please let me know the link to it?

**Comment by tillkruss** [comment]
@nicolas-grekas We haven't published the source of Relay, because of it's commercial tier. We may in the future. As for security, we've done code audits our larger customers in the past with, if that's something you'd like.

**Pull request merged**
